docs/source/tutorials/gateware_usb_device_04.rst

```diff
@@ -0,0 +1,386 @@
+        USBStreamOutEndpoint,
+    )
+    from usb_protocol.types                          import (
+        USBDirection,
+        USBTransferType,
+    )
+
+    VENDOR_ID  = 0x1209 # https://pid.codes/1209/
+    PRODUCT_ID = 0x0001
+
+    MAX_PACKET_SIZE = 512
+
+    class VendorRequestHandler(ControlRequestHandler):
+        ...
+
+    class GatewareUSBDevice(Elaboratable):
+        def create_standard_descriptors(self):
+            descriptors = DeviceDescriptorCollection()
+
+            with descriptors.DeviceDescriptor() as d:
+                d.idVendor           = VENDOR_ID
+                d.idProduct          = PRODUCT_ID
+                d.iManufacturer      = "Cynthion Project"
+                d.iProduct           = "Gateware USB Device"
+                d.bNumConfigurations = 1
+
+            with descriptors.ConfigurationDescriptor() as c:
+                with c.InterfaceDescriptor() as i:
+                    i.bInterfaceNumber = 0
+                    # EP 0x01 OUT - receives bulk data from the host
+                    with i.EndpointDescriptor() as e:
+                        e.bEndpointAddress = USBDirection.OUT.to_endpoint_address(0x01)
+                        e.bmAttributes     = USBTransferType.BULK
+                        e.wMaxPacketSize   = MAX_PACKET_SIZE
+                    # EP 0x82 IN  - transmits bulk data to the host
+                    with i.EndpointDescriptor() as e:
+                        e.bEndpointAddress = USBDirection.IN.to_endpoint_address(0x02)
+                        e.bmAttributes     = USBTransferType.BULK
+                        e.wMaxPacketSize   = MAX_PACKET_SIZE
+
+            return descriptors
+
+        def elaborate(self, platform):
+            ...
+
+This adds two endpoint descriptors to our default interface, each of type ``USBTransferType.BULK`` and with a ``MAX_PACKET_SIZE`` of 512. Where the endpoints differ is in their endpoint address. USB endpoint descriptors encode their direction in an 8 bit endpoint address. The first four bits encode the endpoint number, the next three bits are reserved and set to zero and the final bit encodes the direction; 0 for OUT and 1 for IN.
+
+This means that an OUT endpoint number of 0x01 encodes to an endpoint address of 0x01 while an IN endpoint number of ``0x02`` encodes to the address ``0x82``. (``0b0000_0010 + 0b1000_0000 = 0b1000_0010 = 0x82``)
+
+
+Add USB Stream Endpoints
+------------------------
+
+Once our endpoint descriptors have been added to our device configuration we will need some gateware that will be able to respond to USB requests from the host and allow us to receive and transmit data.
+
+LUNA provides the ``USBStreamOutEndpoint`` and ``USBStreamInEndpoint`` modules which conform to the `Amaranth Data streams <https://amaranth-lang.org/docs/amaranth/latest/stdlib/stream.html>`__ interface. Simply put, streams provide a uniform mechanism for unidirectional exchange of arbitrary data between gateware modules.
```

Comment:
> which conform to the `Amaranth Data streams <https://amaranth-lang.org/docs/amaranth/latest/stdlib/stream.html>`__ interface

This is still not true, right?

Reply:
Well _technically_ they conform as long as you don't mind bundling `first` and `last` in the payload but yes, one of us still has to do the PR for the rest of LUNA to use `amaranth.lib.stream` before this is _strictly_ true. 🥲